File: tests/data_structures/binary_tree/binary_tree.go
// Binary Tree benchmark implementation in Go.
// Tests basic binary search tree operations: insert, search, traverse.

package main

import (
	"fmt"
	"math/rand"
	"time"
)

// TreeNode represents a node in the binary tree
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// BinarySearchTree represents a binary search tree
type BinarySearchTree struct {
	Root *TreeNode
	Size int
}

// NewBinarySearchTree creates a new binary search tree
func NewBinarySearchTree() *BinarySearchTree {
	return &BinarySearchTree{Root: nil, Size: 0}
}

// Insert adds a value to the tree
func (bst *BinarySearchTree) Insert(val int) {
	if bst.Root == nil {
		bst.Root = &TreeNode{Val: val}
		bst.Size++
	} else {
		bst.insertRecursive(bst.Root, val)
	}
}

func (bst *BinarySearchTree) insertRecursive(node *TreeNode, val int) {
	if val < node.Val {
		if node.Left == nil {
			node.Left = &TreeNode{Val: val}
			bst.Size++
		} else {
			bst.insertRecursive(node.Left, val)
		}
	} else if val > node.Val {
		if node.Right == nil {
			node.Right = &TreeNode{Val: val}
			bst.Size++
		} else {
			bst.insertRecursive(node.Right, val)
		}
	}
	// Equal values are ignored (no duplicates)
}

// Search finds a value in the tree
func (bst *BinarySearchTree) Search(val int) bool {
	return bst.searchRecursive(bst.Root, val)
}

func (bst *BinarySearchTree) searchRecursive(node *TreeNode, val int) bool {
	if node == nil {
		return false
	}
	if node.Val == val {
		return true
	} else if val < node.Val {
		return bst.searchRecursive(node.Left, val)
	} else {
		return bst.searchRecursive(node.Right, val)
	}
}

// InorderTraversal performs inorder traversal of the tree
func (bst *BinarySearchTree) InorderTraversal() []int {
	var result []int
	bst.inorderRecursive(bst.Root, &result)
	return result
}

func (bst *BinarySearchTree) inorderRecursive(node *TreeNode, result *[]int) {
	if node != nil {
		bst.inorderRecursive(node.Left, result)
		*result = append(*result, node.Val)
		bst.inorderRecursive(node.Right, result)
	}
}

// GetSize returns the size of the tree
func (bst *BinarySearchTree) GetSize() int {
	return bst.Size
}

// isSorted checks if a slice is sorted
func isSorted(arr []int) bool {
	for i := 1; i < len(arr); i++ {
		if arr[i-1] > arr[i] {
			return false
		}
	}
	return true
}

func main() {
	fmt.Println("Starting binary tree benchmark...")
	startTime := time.Now()
	
	bst := NewBinarySearchTree()
	nodesCount := 1000
	
	// Create shuffled values for insertion
	rand.Seed(42) // For reproducible results
	values := make([]int, nodesCount)
	for i := 0; i < nodesCount; i++ {
		values[i] = i
	}
	rand.Shuffle(len(values), func(i, j int) {
		values[i], values[j] = values[j], values[i]
	})
	
	// Insert operations
	for _, val := range values {
		bst.Insert(val)
	}
	
	// Search operations
	foundCount := 0
	searchCount := 100
	if searchCount > len(values) {
		searchCount = len(values)
	}
	
	for i := 0; i < searchCount; i++ {
		if bst.Search(values[i]) {
			foundCount++
		}
	}
	
	// Traversal operation
	traversalResult := bst.InorderTraversal()
	sorted := isSorted(traversalResult)
	
	executionTime := time.Since(startTime)
	
	fmt.Printf("Tree operations completed: %d inserts, %d searches\n", 
		nodesCount, foundCount)
	fmt.Printf("Final tree size: %d\n", bst.GetSize())
	fmt.Printf("Inorder traversal length: %d\n", len(traversalResult))
	fmt.Printf("Traversal is sorted: %t\n", sorted)
	fmt.Printf("Execution time: %.6f seconds\n", executionTime.Seconds())
}

File: tests/data_structures/binary_tree/binary_tree.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <random>
#include <algorithm>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class BinarySearchTree {
private:
    TreeNode* root;
    int size;
    
    void insertRecursive(TreeNode* node, int val) {
        if (val < node->val) {
            if (node->left == nullptr) {
                node->left = new TreeNode(val);
                size++;
            } else {
                insertRecursive(node->left, val);
            }
        } else if (val > node->val) {
            if (node->right == nullptr) {
                node->right = new TreeNode(val);
                size++;
            } else {
                insertRecursive(node->right, val);
            }
        }
    }
    
    bool searchRecursive(TreeNode* node, int val) {
        if (node == nullptr) return false;
        if (node->val == val) return true;
        if (val < node->val) return searchRecursive(node->left, val);
        return searchRecursive(node->right, val);
    }
    
    void inorderRecursive(TreeNode* node, std::vector<int>& result) {
        if (node != nullptr) {
            inorderRecursive(node->left, result);
            result.push_back(node->val);
            inorderRecursive(node->right, result);
        }
    }
    
    void destroyTree(TreeNode* node) {
        if (node != nullptr) {
            destroyTree(node->left);
            destroyTree(node->right);
            delete node;
        }
    }

public:
    BinarySearchTree() : root(nullptr), size(0) {}
    
    ~BinarySearchTree() {
        destroyTree(root);
    }
    
    void insert(int val) {
        if (root == nullptr) {
            root = new TreeNode(val);
            size++;
        } else {
            insertRecursive(root, val);
        }
    }
    
    bool search(int val) {
        return searchRecursive(root, val);
    }
    
    std::vector<int> inorderTraversal() {
        std::vector<int> result;
        inorderRecursive(root, result);
        return result;
    }
    
    int getSize() const {
        return size;
    }
};

int main() {
    std::cout << "Starting binary tree benchmark..." << std::endl;
    
    auto start = std::chrono::high_resolution_clock::now();
    
    BinarySearchTree bst;
    
    // Insert operations
    const int nodes_count = 1000;
    std::vector<int> values(nodes_count);
    for (int i = 0; i < nodes_count; i++) {
        values[i] = i;
    }
    
    std::random_device rd;
    std::mt19937 g(42); // Fixed seed for reproducible results
    std::shuffle(values.begin(), values.end(), g);
    
    for (int val : values) {
        bst.insert(val);
    }
    
    // Search operations
    int found_count = 0;
    int search_count = std::min(100, (int)values.size());
    
    for (int i = 0; i < search_count; i++) {
        if (bst.search(values[i])) {
            found_count++;
        }
    }
    
    // Traversal operation
    std::vector<int> traversal_result = bst.inorderTraversal();
    bool is_sorted = std::is_sorted(traversal_result.begin(), traversal_result.end());
    
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    double execution_time = duration.count() / 1000000.0;
    
    std::cout << "Tree operations completed: " << nodes_count << " inserts, " 
              << found_count << " searches" << std::endl;
    std::cout << "Final tree size: " << bst.getSize() << std::endl;
    std::cout << "Inorder traversal length: " << traversal_result.size() << std::endl;
    std::cout << "Traversal is sorted: " << (is_sorted ? "true" : "false") << std::endl;
    std::cout << "Execution time: " << std::fixed << execution_time << " seconds" << std::endl;
    
    return 0;
}